### PR TITLE
Fix provider recursion bug, provider function types

### PR DESCRIPTION
**Pull request by christianbaroni**
Fixes APP-1567

## What changed (plus any additional context for devs)
- Fixes an issue where `getProviderForNetwork` was being repeatedly called throughout the app, for every network, due to the `getProvider` function within the network objects being invoked any time `getNetworkObj` or `getXNetworkObj` was invoked
  - Additionally, when `getProviderForNetwork` was called for a not-yet-cached network, the calling of `getNetworkObj` from within `getProviderForNetwork` would then again call `getProviderForNetwork` due to `getProvider` within the network object being invoked, leading to a recursive loop that would run until the provider eventually made it into the cache
- Fixes the return types of the `getProviderForNetwork` and `getCachedProviderForNetwork` functions
- Adds a `getIsHardhatConnected` helper function

### CODE REVIEW

**Review by walmat — APPROVED** [review]

File: src/handlers/web3.ts [hunk 1]
```diff
@@ -207,8 +211,12 @@ export const getFlashbotsProvider = async () => {
...
 
-export const getCachedProviderForNetwork = (network: Network = Network.mainnet) => {
-  return networkProviders[network]!;
+export const getCachedProviderForNetwork = (network: Network = Network.mainnet): StaticJsonRpcProvider | undefined => {
+  const cachedProvider = networkProviders.get(network);
+  if (cachedProvider) {
+    return cachedProvider;
+  }
+  return undefined;
 };
```
Comment (greg-schrammel) [inline src/handlers/web3.ts:220]: ain't this the same? 
```suggestion
export const getCachedProviderForNetwork = (network: Network = Network.mainnet): StaticJsonRpcProvider | undefined => {
  return networkProviders.get(network);
};
```

File: src/handlers/web3.ts [hunk 2]
```diff
@@ -217,27 +225,37 @@ export const getCachedProviderForNetwork = (network: Network = Network.mainnet)
...
-    const provider = new StaticJsonRpcProvider(getNetworkObj(network).rpc, chainId);
-    if (!networkProviders[network]) {
-      networkProviders[network] = provider;
-    }
-    await provider.ready;
+    const provider = new StaticJsonRpcProvider(getNetworkObj(network).rpc(), getNetworkObj(network).id);
+    networkProviders.set(network, provider);
     return provider;
   }
 };
```
Comment (greg-schrammel) [inline src/handlers/web3.ts:244]: the refactor keeps the same behavior, but is it right? if the network is supported but not cached we return `undefined`?

**Review by greg-schrammel — APPROVED** [review]

File: src/handlers/web3.ts (hunk 2 above)
Comment (christianbaroni) [reply src/handlers/web3.ts:244]: If the network is supported but not cached, it'll skip this block:

```typescript
if (isSupportedNetwork && cachedProvider) {
  return cachedProvider;
}
```

And then it'll hit this else block:

```typescript
if (!isSupportedNetwork) {
  // ...
else {
  const provider = new StaticJsonRpcProvider(getNetworkObj(network).rpc(), getNetworkObj(network).id);
  networkProviders.set(network, provider);
  return provider;
}
```

Which creates a new provider and adds it to the cache — where can it return `undefined`?

File: src/handlers/web3.ts (hunk 1 above)
Comment (christianbaroni) [reply src/handlers/web3.ts:220]: Yes good call

**Pull request merged**